server/protocols/jmap/src/main/java/org/apache/james/jmap/UploadServlet.java

```diff
@@ -18,17 +18,78 @@
  ****************************************************************/
 package org.apache.james.jmap;
 
+import static javax.servlet.http.HttpServletResponse.SC_BAD_REQUEST;
 import static javax.servlet.http.HttpServletResponse.SC_CREATED;
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
 
+import java.io.IOException;
+import java.io.InputStream;
+
+import javax.inject.Inject;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+import org.apache.james.jmap.model.UploadResponse;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.mail.AttachmentMapper;
+import org.apache.james.mailbox.store.mail.model.Attachment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Strings;
+import com.google.common.io.ByteStreams;
+
 public class UploadServlet extends HttpServlet {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UploadServlet.class);
+
+    private final MailboxSessionMapperFactory mailboxSessionMapperFactory;
+    private final ObjectMapper objectMapper;
+
+    @Inject
+    private UploadServlet(MailboxSessionMapperFactory mailboxSessionMapperFactory) {
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+        this.objectMapper = new ObjectMapper();
+    }
 
     @Override
     protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException {
-        resp.setStatus(SC_CREATED);
+        String contentType = req.getContentType();
+        if (Strings.isNullOrEmpty(contentType)) {
+            resp.setStatus(SC_BAD_REQUEST);
+        } else {
+            try {
+                UploadResponse storedContent = uploadContent(contentType, req.getInputStream());
```

Comment:
uploadContent and buildResponse belongs to another layer, the servlet should only handle technical stuff.